### PR TITLE
Added loading spinner for search requests

### PR DESCRIPTION
**Pull request by ryanlieu**
### Summary of Changes

Title, we needed some form of loading state so that delays between search requests wouldn't look awkward. There's a small hack here where we need to handle GetPopularTables actions from the Search reducer but this'll be addressed in a follow up with the decoupling of popular tables and search in the new home page

### Tests

N/A

### Documentation

N/A

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes, including screenshots of any UI changes. 
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public python functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes all tests documented in the [developer guide](https://github.com/lyft/amundsenfrontendlibrary/blob/master/docs/developer_guide.md#testing)
- [ ] I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"

### CODE REVIEW

**Comment by ttannis** [comment]
👀

File: amundsen_application/static/js/ducks/search/reducer.ts [hunk 1]
```diff
@@ -64,25 +72,39 @@ export default function reducer(state: SearchReducerState = initialState, action
       return {
         ...state,
         search_term: action.term,
+        isLoading: true,
```
Comment (danwom) [inline amundsen_application/static/js/ducks/search/reducer.ts]: Looks like only the `SearchAll` action will trigger the loading spinner. Do you also want to show the spinner for a `SearchResource` action as well? Generally this will happen when a user clicks on a paginated result. It may or may not be a good user experience.

File: amundsen_application/static/js/components/SearchPage/index.tsx [hunk 2]
```diff
@@ -250,15 +252,24 @@ export class SearchPage extends React.Component<SearchPageProps, SearchPageState
       );
   };
 
+  renderTabs = () => {
```
Comment (danwom) [inline amundsen_application/static/js/components/SearchPage/index.tsx]: I think `renderContent` or `renderSearchResults` would be better names, since not every section here has tabs. This will also be where the default homepage content will be rendered as well.

File: amundsen_application/static/js/components/SearchPage/index.tsx [hunk 3]
```diff
@@ -8,6 +8,7 @@ import { RouteComponentProps } from 'react-router';
 
 import SearchBar from './SearchBar';
 import SearchList from './SearchList';
+import LoadingSpinner from 'components/common/LoadingSpinner';
```
Comment (danwom) [inline amundsen_application/static/js/components/SearchPage/index.tsx]: This might be out of scope, but the LoadingSpinner has a fixed height and width that may not look great in all places. Maybe look into how to make it smaller/bigger depending on the context.

File: amundsen_application/static/js/components/SearchPage/index.tsx (hunk 2 above)
Comment (ttannis) [reply amundsen_application/static/js/components/SearchPage/index.tsx]: +1 for `renderContent`

File: amundsen_application/static/js/ducks/search/reducer.ts [hunk 4]
```diff
@@ -64,25 +72,39 @@ export default function reducer(state: SearchReducerState = initialState, action
...
+      };      
+    case GetPopularTables.SUCCESS:
+    case GetPopularTables.FAILURE:
+      return {
+        ...state,
+        isLoading: false,
+      };
     default:
-      return state;
+      return {
```
Comment (ttannis) [inline amundsen_application/static/js/ducks/search/reducer.ts]: Why the change?

File: amundsen_application/static/js/ducks/search/reducer.ts [hunk 5]
```diff
@@ -40,6 +47,7 @@ export function searchResource(resource: ResourceType, term: string, pageIndex:
 
 const initialState: SearchReducerState = {
   search_term: '',
+  isLoading: true,
```
Comment (ttannis) [inline amundsen_application/static/js/ducks/search/reducer.ts]: Can we default to `false`, or do you currently end up seeing the no results message followed by the spinner followed by the actual search results when it completes? 

Would that also remove the need to have the following code?:
```
 case GetPopularTables.SUCCESS:
 case GetPopularTables.FAILURE:
      return {
        ...state,
        isLoading: false,
      };
```

**Comment by danwom** [comment]
I just realized this about your previous PR but when a user uses the pagination controls, `onPaginationChange` fetches the new page data with `this.props.searchResource`, and then updates the page URL with `this.updatePageUrl`. The URL update then triggers `componentDidUpdate` which then calls `this.props.searchAll`. So this actually makes two requests for the same data.


This isn't a big deal now, but this might break some behavior integrations with multiple search resource types. (Someone recently implemented Dashboards)

File: amundsen_application/static/js/fixtures/globalState.ts [hunk 6]
```diff
@@ -41,6 +41,7 @@ const globalState: GlobalState = {
   ],
   search: {
     search_term: 'testName',
+    isLoading: true,
```
Comment (ttannis) [inline amundsen_application/static/js/fixtures/globalState.ts]: Change to false if you end up making the default value `false`.

File: amundsen_application/static/js/ducks/search/reducer.ts (hunk 1 above)
Comment (ttannis) [reply amundsen_application/static/js/ducks/search/reducer.ts]: Similar comment here, I see you reset the loading state when `SearchResource` completes/fails but didn't set the loading state to true when `SearchResource` is called. I'd be interested to checkout what this looks and feels like when we use the LoadingSpinner for both cases.

File: amundsen_application/static/js/components/SearchPage/index.tsx (hunk 3 above)
Comment (ryanlieu) [reply amundsen_application/static/js/components/SearchPage/index.tsx]: I think this is probably out of scope for this PR, will file a follow up task

File: amundsen_application/static/js/components/SearchPage/index.tsx (hunk 2 above)
Comment (ryanlieu) [reply amundsen_application/static/js/components/SearchPage/index.tsx]: :+1:

File: amundsen_application/static/js/ducks/search/reducer.ts (hunk 4 above)
Comment (ryanlieu) [reply amundsen_application/static/js/ducks/search/reducer.ts]: oops, leftover code from some testing 😅

**Comment by ryanlieu** [comment]
@danwom I'm going to lumping in a fix to that multiple request bug to this MR as it was causing the spinner to show up even on pagination (which wasn't intended).

File: amundsen_application/static/js/ducks/search/reducer.ts (hunk 5 above)
Comment (ryanlieu) [reply amundsen_application/static/js/ducks/search/reducer.ts]: just tried it out, that works and is way cleaner!

File: amundsen_application/static/js/ducks/search/reducer.ts (hunk 1 above)
Comment (ryanlieu) [reply amundsen_application/static/js/ducks/search/reducer.ts]: updating so searchresource action also sets isLoading to true

File: amundsen_application/static/js/ducks/search/reducer.ts [hunk 7]
```diff
@@ -61,27 +63,34 @@ export default function reducer(state: SearchReducerState = initialState, action
   switch (action.type) {
     // Updates search term to reflect action
     case SearchAll.ACTION:
+    case SearchResource.ACTION:
```
Comment (ttannis) [inline amundsen_application/static/js/ducks/search/reducer.ts]: Because searching a resource doesn't change the search_term, there's no need to update it here. We can keep out logic cleaner and more intentional by instead having:
```
case SearchResource.ACTION:
    return {
        ...state,
        isLoading: true,
     }
```

**Comment by ryanlieu** [comment]
👀 @ttannis @danwom this should be gtg with tests

File: amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx [hunk 8]
```diff
@@ -273,13 +270,47 @@ describe('SearchPage', () => {
       wrapper.instance().componentDidUpdate(mockPrevProps);
     });
 
-    it('calls setState', () => {
-      expect(setStateSpy).toHaveBeenCalledWith({ selectedTab: ResourceType.table });
-    });
+    describe('called with a new search term', () => {
+      it('calls setState', () => {
```
Comment (ttannis) [inline amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx]: We call `setState` regardless of what happens with the search term so it's misleading to bundle that into this case. These two checks can be separate tests and don't have to be bundled into the same block:
```
it('calls setState', () => {
    expect(setStateSpy).toHaveBeenCalledWith({ selectedTab: ResourceType.table });
});

it('calls searchAll if called with a new search term', () => {
   expect(searchAllSpy).toHaveBeenCalledWith(mockSanitizedUrlParams.term, mockSearchOptions);
});
```
 
However one thing we can keep is the fact that you added some more explicit titles for the tests, to help clarify the conditional logic that we're testing.

File: amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx [hunk 9]
```diff
@@ -273,13 +270,47 @@ describe('SearchPage', () => {
...
+      });
 
-    it('calls searchAll', () => {
-      expect(searchAllSpy).toHaveBeenCalledWith(mockSanitizedUrlParams.term, mockSearchOptions);
+      it('calls searchAll', () => {
+        expect(searchAllSpy).toHaveBeenCalledWith(mockSanitizedUrlParams.term, mockSearchOptions);
+      });
     });
+
+    describe('called with a new page but with the same search term', () => {
```
Comment (ttannis) [inline amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx]: We can condense this by leveraging the setup that was used initially for these tests. This is because this test case doesn't necessarily need a new setup, we only have to call `wrapper.instance().componentDidUpdate()` with some new `mockPrevProps` and verify what happens with the `searchAllSpy`.
```
it('does not call searchAll if called with the same search term', () => {
    /* assuming we use the same setup from the previous tests and don't have to write all the boiler plate again */
    searchAllSpy.mockClear(); // because another test will have called it.
     const mockPrevProps = {
          location: {
            search: '/search?searchTerm=current&current=table&pageIndex=1', 
            pathname: 'mockstr',
            state: jest.fn(),
            hash: 'mockstr',
          }
     };
    wrapper.instance().componentDidUpdate(mockPrevProps);
    expect(searchAllSpy).not.toHaveBeenCalled();
});
```

The above should work, but let me know if you run into any issues.

File: amundsen_application/static/js/components/SearchPage/index.tsx [hunk 10]
```diff
@@ -96,7 +98,10 @@ export class SearchPage extends React.Component<SearchPageProps, SearchPageState
       const { searchTerm, pageIndex, selectedTab } = params;
       const { term, index, currentTab } = this.getSanitizedUrlParams(searchTerm, pageIndex, selectedTab);
       this.setState({ selectedTab: currentTab });
-      this.props.searchAll(term, this.createSearchOptions(index, currentTab));
+      const prevTerm = qs.parse(prevProps.location.search).searchTerm;
```
Comment (danwom) [inline amundsen_application/static/js/components/SearchPage/index.tsx]: Rather than re-parsing the previous props location, you can pull in the prev state and use that value. Interface for this function is:

`componentDidUpdate(prevProps, prevState, snapshot)`

File: amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx [hunk 11]
```diff
@@ -606,6 +637,50 @@ describe('SearchPage', () => {
     });
   });
 
+  describe('renderContent', () => {
+    let content;
```
Comment (ttannis) [inline amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx]: I see you're not actually using the `content` variable in your tests so we can remove those.

File: amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx [hunk 12]
```diff
@@ -606,6 +637,50 @@ describe('SearchPage', () => {
...
+    describe('popular tables', () => {
+      beforeAll(() => {
+        const setupResult = setup({ searchTerm: '' });
+        props = setupResult.props;
+        wrapper = setupResult.wrapper;
+        content = shallow(wrapper.instance().renderContent());
+      });
+      
+      it('renders popular tables', () => {
+        expect(wrapper.instance().renderContent()).toEqual(wrapper.instance().renderPopularTables());
```
Comment (ttannis) [inline amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx]: nit: Another approach would be to spy on `renderPopularTables` and `renderSearchResults` and check that they were called. Imagine if we had thousands of tests with several hundred of them doing similar checks -- spying on methods whenever we can is expected be faster than if those hundreds of tests were rendering content and checking the output each time. Fine to merge as is though if you like.

File: amundsen_application/static/js/components/SearchPage/index.tsx (hunk 10 above)
Comment (ryanlieu) [reply amundsen_application/static/js/components/SearchPage/index.tsx]: oh right, I could just set it prevProps.searchTerm right?

Comment (danwom) [reply amundsen_application/static/js/components/SearchPage/index.tsx]: There's an important distinction between `state` and `props`. With the way we're using `Redux/Saga` the 'container' pulls values from the `globalState` and sends it as `props` on the `component`. 

Sometimes the component can use the props as-is, but other times it needs to consume props to update its internal `state`.

File: amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx [hunk 13]
```diff
@@ -606,6 +637,50 @@ describe('SearchPage', () => {
     });
   });
 
+  describe('renderContent', () => {
+    let content;
+    let props;
+    let wrapper;
+    describe('popular tables', () => {
+      beforeAll(() => {
```
Comment (ttannis) [inline amundsen_application/static/js/components/SearchPage/tests/index.spec.tsx]: nit: Since each we only have one case for each test, we don't necessarily need a new `describe` block and a `beforeAll`, we could also have done:

```
it('renders popular tables if searchTerm is empty', () => {
    const setupResult = setup({ searchTerm: '' });
    props = setupResult.props;
    wrapper = setupResult.wrapper;
    expect(wrapper.instance().renderContent()).toEqual(wrapper.instance().renderPopularTables());
});
```
Note the addition of clarifying the conditional logic we're testing in the description as well.

**Comment by ryanlieu** [comment]
👀 @ttannis @danwom addressed comments, should be good to go.
